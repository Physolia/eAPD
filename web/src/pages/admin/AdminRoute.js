import PropTypes from 'prop-types';
import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { getIsFedAdmin } from '../../reducers/user.selector';
import NoMatch from '../../components/NoMatch';

const AdminRoute = ({
  authenticated,
  component: Component,
  isAdmin,
  ...rest
}) => {
  return (
    <Route
      {...rest}
      render={props => {
        if (authenticated) {
          if (isAdmin) {
            return <Component {...props} />;
          }

          // If they're logged in but not an admin, display the equivalent of
          // a 404 page rather than reveal this page exists.
          return <NoMatch />;
        }
        return (
          <Redirect
            to={{ pathname: '/login', state: { from: props.location } }}
          />
        );
      }}
    />
  );
};

AdminRoute.propTypes = {
  authenticated: PropTypes.bool.isRequired,
  component: PropTypes.func.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  location: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  authenticated: state.auth.authenticated,
  isAdmin: getIsFedAdmin(state)
});

export default connect(mapStateToProps)(AdminRoute);

export { AdminRoute as plain, mapStateToProps };
