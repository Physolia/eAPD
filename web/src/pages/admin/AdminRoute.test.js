import { shallow } from 'enzyme';
import React from 'react';

import { plain as AdminRoute, mapStateToProps } from './AdminRoute';

describe('admin route component', () => {
  const props = {
    authenticated: false,
    isAdmin: false,
    location: { from: 'over there' },
    component: () => {}
  };

  test('renders correctly if logged out', () => {
    const component = shallow(<AdminRoute {...props} />);
    expect(component).toMatchSnapshot();
    expect(component.prop('render')(props)).toMatchSnapshot();
  });

  test('renders correctly if logged in but not admin', () => {
    const component = shallow(<AdminRoute {...props} authenticated />);
    expect(component).toMatchSnapshot();
    expect(
      component.prop('render')({ ...props, authenticated: true })
    ).toMatchSnapshot();
  });

  test('renders correctly if logged in and admin', () => {
    const component = shallow(<AdminRoute {...props} authenticated isAdmin />);
    expect(component).toMatchSnapshot();
    expect(
      component.prop('render')({ ...props, authenticated: true, isAdmin: true })
    ).toMatchSnapshot();
  });

  test('maps state to props', () => {
    expect(
      mapStateToProps({
        auth: {
          authenticated: 'some value'
        },
        user: { data: { role: 'not admin' } }
      })
    ).toEqual({ authenticated: 'some value', isAdmin: false });

    expect(
      mapStateToProps({
        auth: {
          authenticated: 'some value'
        },
        user: { data: { role: 'eAPD Federal Admin' } }
      })
    ).toEqual({ authenticated: 'some value', isAdmin: true });
  });
});
