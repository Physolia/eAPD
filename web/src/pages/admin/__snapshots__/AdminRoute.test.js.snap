// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin route component renders correctly if logged in and admin 1`] = `
<Route
  location={
    Object {
      "from": "over there",
    }
  }
  render={[Function]}
/>
`;

exports[`admin route component renders correctly if logged in and admin 2`] = `
<component
  authenticated={true}
  component={[Function]}
  isAdmin={true}
  location={
    Object {
      "from": "over there",
    }
  }
/>
`;

exports[`admin route component renders correctly if logged in but not admin 1`] = `
<Route
  location={
    Object {
      "from": "over there",
    }
  }
  render={[Function]}
/>
`;

exports[`admin route component renders correctly if logged in but not admin 2`] = `<NoMatch />`;

exports[`admin route component renders correctly if logged out 1`] = `
<Route
  location={
    Object {
      "from": "over there",
    }
  }
  render={[Function]}
/>
`;

exports[`admin route component renders correctly if logged out 2`] = `
<Redirect
  to={
    Object {
      "pathname": "/login",
      "state": Object {
        "from": Object {
          "from": "over there",
        },
      },
    }
  }
/>
`;
